phoenix-core/src/main/java/org/apache/phoenix/iterate/BaseResultIterators.java

```diff
@@ -491,6 +476,22 @@ private static void optimizeProjection(StatementContext context, Scan scan, PTab
                 }
             }
         }
+        for (Entry<byte[], NavigableSet<byte[]>> entry : familyMap.entrySet()) {
+            ImmutableBytesPtr cf = new ImmutableBytesPtr(entry.getKey());
+            NavigableSet<byte[]> qs = entry.getValue();
+            NavigableSet<ImmutableBytesPtr> cols = null;
+            if (qs != null) {
+                cols = new TreeSet<ImmutableBytesPtr>();
+                for (byte[] q : qs) {
+                    cols.add(new ImmutableBytesPtr(q));
+                    if (trackedColumnsBitset != null) {
+                        int qualifier = encodingScheme.decode(q);
+                        trackedColumnsBitset.set(qualifier);
+                    }
+                }
+            }
+            columnsTracker.put(cf, cols);
+        }
```

Comment:
I think the side effect of moving this code block here would be adding some extra columns to the result set. I do not think that would create a functional issue but just result in more data to return to the client.

Reply:
I think the extra columns would also only be added of they are needed. Without this the construction of the Filters below would work on a different set of columns as compared to the scan objects. It seems we were lucky that we did not run into this.
This does not add more columns to the scan, just to the column tracker and the extra bits to the trackedColumnsBitSet used in the filters below.
But I agree... In the worst case we added extra columns to the tracker(s).